### PR TITLE
aci_l3out: Add initial ACI l3Out implementation

### LINKED ISSUE #37457: Need module for L3Out configuration (basic + adding external EPGs)
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME

aci_aep
##### ANSIBLE VERSION

```
ansible 2.4.3.0
```

##### SUMMARY
Need module for L3Out configuration (basic + adding external EPGs)

##### STEPS TO REPRODUCE

```yaml

```

##### ACTUAL RESULTS

```

```

### PR DESCRIPTION
**Pull request by rost-d**
##### SUMMARY
Implementation of ACI L3Out module

Tested on 3.1(1i)

This fixes #37457 

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
aci_l3out

##### ANSIBLE VERSION
```
ansible 2.4.3.0
  config file = None
  configured module search path = ['/Users/rdavyden/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python3.6/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 3.6.4 (default, Mar  1 2018, 18:36:50) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

### CODE REVIEW

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+def main():
+    argument_spec = aci_argument_spec()
+    argument_spec.update(
+        l3out=dict(type='str', aliases=['l3out_name', 'name']),
+        domain=dict(type='str', aliases=[
+                    'ext_routed_domain_name', 'routed_domain']),
+        vrf=dict(type='str', aliases=['vrf_name']),
+        tenant=dict(type='str', aliases=['tenant_name']),
+        description=dict(type='str', aliases=['descr']),
+        route_control=dict(type='str', default='export', choices=[
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: This ought to become:
```python
    route_control=dict(type='list', default=['export'], choices=['export', 'import'], aliases=['route_control_enforcement']),
```
Then check whether the value it not None, and does not equal ['import']. The parameter description then would state that the value is either export, or [ export, import ]

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+                           'export,import', 'export'], aliases=['route_control_enforcement']),
+        dscp=dict(type='str',
+                  choices=['AF11', 'AF12', 'AF13', 'AF21', 'AF22', 'AF23', 'AF31', 'AF32', 'AF33', 'AF41', 'AF42', 'AF43',
+                           'CS0', 'CS1', 'CS2', 'CS3', 'CS4', 'CS5', 'CS6', 'CS7', 'EF', 'VA', 'unspecified'],
+                  aliases=['target']),
+        l3protocol=dict(type='str', default='static',
+                        choices=['static', 'bgp', 'ospf']),
+        state=dict(type='str', default='present', choices=[
+                   'absent', 'present', 'query']),
+        method=dict(type='str', choices=['delete', 'get', 'post'], aliases=[
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Don't add this, we are deprecating this for modules that predate v2.4.
So very soon all modules will have this removed.

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+                    'action'], removed_in_version='2.6'),  # Deprecated starting from v2.6
+        # Deprecated in v2.6
+        protocol=dict(type='str', removed_in_version='2.6'),
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Don't add this, we are deprecating this for modules that predate v2.4.

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+    child_configs = [
+        dict(l3extRsL3DomAtt=dict(attributes=dict(
+            tDn='uni/l3dom-{0}'.format(domain)))),
+        dict(l3extRsEctx=dict(attributes=dict(tnFvCtxName=vrf))),
+    ]
+    if l3protocol == 'bgp':
+        child_configs.append(
+            dict(bgpExtP=dict(attributes=dict(descr='', nameAlias=''))))
+    if l3protocol == 'ospf':
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Make this `elif`.

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+    if state == 'present':
+        aci.payload(
+            aci_class='l3extOut',
+            class_config=dict(
+                name=l3out,
+                descr=description,
+                dn='uni/tn-{0}/out-{1}'.format(tenant, l3out),
+                enforceRtctrl=enforceRtctrl,
+                targetDscp=dscp,
+                nameAlias='',
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Why clear these explicitly ?

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+                ownerKey='',
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Why clear these explicitly ?

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+                ownerTag='',
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Why clear these explicitly ?

**Review by dagwieers — CHANGES_REQUESTED** [review]
Looking good, but we need wide feedback about potential limitations of the module's user interface and needed functionality. And we have the time to get this right by v2.6... ;-)

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    - Description for the L3Out.
+    default:
+    aliases: [ descr ]
+  state:
+    description:
+    - Use C(present) or C(absent) for adding or removing.
+    - Use C(query) for listing an object or multiple objects.
+    choices: [ absent, present, query ]
+    default: present
+  method:
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py]: Please also remove this. Legacy ;-)

**Review by dagwieers — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/network/aci/aci_l3out.py (hunk 1 above)
Comment (jmcgill298) [reply lib/ansible/modules/network/aci/aci_l3out.py]: All configuration values should not include default values, as that leads to accidental overwriting of existing configs that the end user did not expect.

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 9]
```diff
@@ -0,0 +1,313 @@
...
+  dscp:
+    description:
+    - The target Differentiated Service (DSCP) value.
+    choices: [ AF11, AF12, AF13, AF21, AF22, AF23, AF31, AF32, AF33, AF41, AF42, AF43, CS0, CS1, CS2, CS3, CS4, CS5, CS6, CS7, EF, VA, unspecified ]
+    default: unspecified
+    aliases: [ target ]
+  route_control:
+    description:
+    - Route Control enforcement direction. The only allowed values are export or import,export.
+    choices: [ 'export', 'import' ]
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py:57]: In this case quotes are not required, but it is not a sin ;-)

File: lib/ansible/modules/network/aci/aci_l3out.py [hunk 10]
```diff
@@ -0,0 +1,313 @@
...
+    aliases: [ descr ]
+  state:
+    description:
+    - Use C(present) or C(absent) for adding or removing.
+    choices: [ absent, present ]
+    default: present
+extends_documentation_fragment: aci
+'''
+
+EXAMPLES = r'''
```
Comment (dagwieers) [inline lib/ansible/modules/network/aci/aci_l3out.py:77]: We prefer also examples where we remove objects, query a single object or query all objects. Look at some of the other modules. We have at minimum 4 examples.

**Review by dagwieers — CHANGES_REQUESTED** [review]

**Comment by dagwieers** [comment]
cc @rsmeyers @nilloBE

**Review by dagwieers — APPROVED** [review]
LGTM

File: lib/ansible/modules/network/aci/aci_l3out.py (hunk 10 above)
Comment (dagwieers) [reply lib/ansible/modules/network/aci/aci_l3out.py:77]: Thanks !

**Comment by dagwieers** [comment]
Can we fix the reported CI issue ?

The DSCP value "unspecified" should either be handled, or removed (and assumed null/none).

**Comment by dagwieers** [comment]
I am fixing the other 2 issues, as soon as that is merged I will merge your module.
And it will ship with v2.6.

Next to this, we need to create all the additional functionality people would need for a properly working l3out, and that's a *lot* of work...

**Review by jmcgill298 — CHANGES_REQUESTED** [review]
configuration values should not have a default value, as that will cause the users to implement unintended configuration changes (dscp is set to AF11, and not sent as a parameter in the task; now the dscp is 'unspecified'). The APIC will handle the default for us if the object does not exist, and the diffs ignore all unpassed parameters.

**Comment by dagwieers** [comment]
@jmcgill298 Correct, that is how I am fixing this now to accommodate the new module validation checks.
```diff
   dscp:
     description:
     - The target Differentiated Service (DSCP) value.
+    - The APIC defaults to C(unspecified) when unset during creation.
     choices: [ AF11, AF12, AF13, AF21, AF22, AF23, AF31, AF32, AF33, AF41, AF42, AF43, CS0, CS1, CS2, CS3, CS4, CS5, CS6, CS7, EF, VA, unspecified ]
-    default: unspecified
     aliases: [ target ]
```

PR is almost ready, will ask you to review it :)

**Comment by rost-d** [comment]
So we need to remove default for DSCP from documentation I guess... We have a lot of modules mentioning unspecified as a default value for this parameter.

aci_l3out_node is almost ready (Node profile - tested but needs some documentation update)

**Comment by dagwieers** [comment]
@rost-d I am fixing all of them, hold on :-)

**Comment by dagwieers** [comment]
@rost-d If you plan to become an active ACI contributor, welcome !

I can add you to the ACI team so you can join future meetings, contribute to the next roadmap (v2.7) , and review/merge ACI-related PRs ! We are collaborating in the Wiki at: https://github.com/ansible/community/wiki/Network:-ACI

**Comment by dagwieers** [comment]
@rost-d Thanks for your contribution ! Wiki updated...

**Comment by brunocalogero** [comment]
nice! @rost-d : -)

**Pull request merged**